src/typ.rs

```diff
@@ -2205,6 +2216,13 @@ fn infer_let(
     let pattern = PatternTyper::new(env, level).unify(pattern, value_typ)?;
     let then = infer(then, level, env)?;
     let typ = then.typ();
+
+    // Check that any type annotation is accurate.
+    if let Some(ann) = annotation {
+        let ann_typ = env.type_from_ast(ann, &mut hashmap![], NewTypeAction::MakeGeneric)?;
+        unify(ann_typ, value.typ(), env).map_err(|e| convert_unify_error(e, value.location()))?;
```

Comment:
Similarly, is it correct for me to use `value.typ()` directly, without generalising it first?